### ISSUE TITLE
Web components with ShadowDOM, unable to focus on elements in slot (LightDOM)

### ISSUE THREAD

**Issue by kee-oth** [focus-trap/focus-trap#601]
I'm using `focus-trap` in a web components library but I am unable to make the package work with elements within a ShadowDOM `slot`. Is there a way to have a single focus trap contain elements from both the ShadowDOM and elements passed in via the `slot`?

Example:
```
// modal.js

const Modal = (props) => {

  const trapFocus = (element) => {
    const trap = focusTrap.createFocusTrap(element)
    trap.activate()
  }

  return (
    <div ref={trapFocus}>
      <a href="#">These items</a>
      <a href="#">get trapped</a>
      <slot /> // Elements in slot don't get trapped
    </div>
  );
}

// index.html
<custom-modal>
    <button>This goes in the slot and isn't included in the focus trap</button> 
</custom-modal>
```

Thank you!

**Comment by stefcameron** [comment]
@kee-oth Thank you for your question. Actually, work is pending on some level of Shadow DOM support over in https://github.com/focus-trap/tabbable/pull/304, including support for slots. Do you think the solution there looks like it might solve your issue?

**Comment by kee-oth** [comment]
@stefcameron Thanks for the reply! It does indeed look like the functionality that I need is being addressed in that thread 👍

It seems like you're actively trying to get that functionality in which I appreciate. I'll keep tabs on that thread for now, thank you!

**Comment by kee-oth** [comment]
@stefcameron And just to point out, this library does work in the situation I need. It's just not really maintained and I'd like to switch my projects to use your package since it is clearly maintained and has a larger community:

https://www.npmjs.com/package/@a11y/focus-trap

Thanks!

**Comment by stefcameron** [comment]
Great! Hopefully I can get a response from @idoros over on https://github.com/focus-trap/tabbable/pull/304 about https://github.com/focus-trap/tabbable/pull/304#issuecomment-986820582 and then I should be able to get a beta out of tabbable and will probably need a beta of focus-trap (this one) as well, and you can give it a try.

**Comment by kee-oth** [comment]
@stefcameron Thanks so much! 🎉

**Comment by stefcameron** [comment]
OK, `focus-trap@6.8.0-beta.0` is published, which includes `tabbable@5.3.0-beta.0`, which includes the new shadow DOM features in https://github.com/focus-trap/tabbable/pull/304 that are now in https://github.com/focus-trap/tabbable/pull/570

Please give it a try and let me know how it goes. I really hope it "just works" for you. 😅

**Comment by stefcameron** [comment]
@kee-oth Just wondering if you noticed I published the new feature in betas of tabbable and focus-trap ☝️ and whether they fix this issue or not.

**Comment by stefcameron** [comment]
@andresriveratoro Nice, glad to hear you were successful with it!

**Comment by kee-oth** [comment]
@stefcameron Thanks for much for the work on this! I'll give this a try shortly 👍

**Comment by andresriveratoro** [comment]
@stefcameron I'm not sure if I'm doing something wrong, but I am unable to complete the loop using `shift+tab`.

I can use `tab` indefinitely to loop through all the focusable elements, but `shift+tab` gets stuck in the 2 containers at the bottom as seen here. Any ideas?


https://user-images.githubusercontent.com/36863582/156242360-fbc7585e-716a-4636-abfc-bcc83993553e.mov

**Comment by andresriveratoro** [comment]
I ended up fixing this by wrapping the first button in its own div and updating the focus trap. I hope this ends up helping someone. Thanks again, @stefcameron!

```jsx
this.focusTrap = createFocusTrap([this.el, this.alertActions, this.closeButtonContainer]);

<div>
 <div ref={el => (this.closeButtonContainer = el as HTMLDivElement)}>
  <button>X</button>
 </div>
 <div>
   <div>
     <slot></slot>
   </div>
   <div ref={el => (this.alertActions = el as HTMLDivElement)}>
     <button>Primary Button</button>
     <button>Secondary Button</button>
   </div>
 </div>
</div>

```

**Comment by stefcameron** [comment]
@andresriveratoro That's interesting, thank you!

So if I'm reading your code correctly, `this.el` is the top-level wrapping `<div>` that contains 2 children, which are referenced by `this.closeButtonContainer` and `this.alertActions`.

And I think you mentioned that `this.closeButtonContainer` and `this.alertActions` are in an __open__ shadow DOM.

I'm glad things are working, but I find it a bit strange that you have to provide the parent and 2 children (of that same parent, that seem to be open shadow DOM roots) as separate focus-trap containers to define the trap.

You _should_ be able to get this to work just with `this.focusTrap = createFocusTrap(this.el)`, but I'm waiting for a bit of help here to confirm: https://github.com/focus-trap/tabbable/pull/570#issuecomment-1057475264

**Comment by Mottie** [comment]
Are there any plans on releasing 6.8.0 soon?

**Comment by stefcameron** [comment]
Probably in a couple of weeks. The folks who were checking out the beta (here and on tabbable) have gone quiet for a while now, not sure if that's because they haven't had time, lost interest, or everything is working well so, "no news is good news."

**Comment by stefcameron** [comment]
It's been quiet for a while. I've decided to move forward. I just published tabbable v5.3.0 with the shadow DOM support.

I think the original issue brought up here was either worked around, or is solved in the v5.3.0. If I'm mistaken, please re-open and we'll try to figure it out.

I will be publishing focus-trap v6.8.0 now.
